test/e2e_node/memory_eviction_test.go

```diff
@@ -0,0 +1,191 @@
+
+	Context("When there is memory pressure", func() {
+		It("It should evict pods in the correct order (besteffort first, then burstable, then guaranteed)", func() {
+			By("Creating a guaranteed pod, a burstable pod, and a besteffort pod.")
+
+			// A pod is guaranteed only when requests and limits are specified for all the containers and they are equal.
+			guaranteed := createMemhogPod(f, "guaranteed-", "guaranteed", api.ResourceRequirements{
+				Requests: api.ResourceList{
+					"cpu":    resource.MustParse("100m"),
+					"memory": resource.MustParse("100Mi"),
+				},
+				Limits: api.ResourceList{
+					"cpu":    resource.MustParse("100m"),
+					"memory": resource.MustParse("100Mi"),
+				}})
+
+			// A pod is burstable if limits and requests do not match across all containers.
+			burstable := createMemhogPod(f, "burstable-", "burstable", api.ResourceRequirements{
+				Requests: api.ResourceList{
+					"cpu":    resource.MustParse("100m"),
+					"memory": resource.MustParse("100Mi"),
+				}})
+
+			// A pod is besteffort if none of its containers have specified any requests or limits.
+			besteffort := createMemhogPod(f, "besteffort-", "besteffort", api.ResourceRequirements{})
+
+			// We poll until timeout or all pods are killed.
+			// Inside the func, we check that all pods are in a valid phase with
+			// respect to the eviction order of best effort, then burstable, then guaranteed.
+			By("Polling the Status.Phase of each pod and checking for violations of the eviction order.")
+			Eventually(func() bool {
+
+				gteed, gtErr := f.Client.Pods(f.Namespace.Name).Get(guaranteed.Name)
+				framework.ExpectNoError(gtErr, fmt.Sprintf("getting pod %s", guaranteed.Name))
+				gteedPh := gteed.Status.Phase
+
+				burst, buErr := f.Client.Pods(f.Namespace.Name).Get(burstable.Name)
+				framework.ExpectNoError(buErr, fmt.Sprintf("getting pod %s", burstable.Name))
+				burstPh := burst.Status.Phase
+
+				best, beErr := f.Client.Pods(f.Namespace.Name).Get(besteffort.Name)
+				framework.ExpectNoError(beErr, fmt.Sprintf("getting pod %s", besteffort.Name))
+				bestPh := best.Status.Phase
+
+				glog.Infof("Pod phase: guaranteed: %v, burstable: %v, besteffort: %v", gteedPh, burstPh, bestPh)
+
+				if bestPh == api.PodRunning {
+					Expect(burstPh).NotTo(Equal(api.PodFailed), "Burstable pod failed before best effort pod")
+					Expect(gteedPh).NotTo(Equal(api.PodFailed), "Guaranteed pod failed before best effort pod")
+				} else if burstPh == api.PodRunning {
+					Expect(gteedPh).NotTo(Equal(api.PodFailed), "Guaranteed pod failed before burstable pod")
+				}
+
+				// When both besteffort and burstable have been evicted, return true, else false
+				if bestPh == api.PodFailed && burstPh == api.PodFailed {
+					return true
+				}
+				return false
+
+			}, 20*time.Minute, 5*time.Second).Should(Equal(true))
+
+		})
+	})
+
+})
+
+func createMemhogPod(f *framework.Framework, genName string, ctnName string, res api.ResourceRequirements) *api.Pod {
```

Comment:
On that note, I created vish/stress because memhog was a large image and it does not let us control the rate of consumption of memory. 

$ docker run --rm vish/stress --help
Usage of /stress:
  -alsologtostderr
    	log to standard error as well as files
  -cpus int
    	total number of CPUs to utilize
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace (default :0)
  -log_dir string
    	If non-empty, write log files in this directory
  -logtostderr
    	log to standard error instead of files
  -mem-alloc-size string
    	amount of memory to be consumed in each allocation (default "4Ki")
  -mem-alloc-sleep duration
    	duration to sleep between allocations (default 1ms)
  -mem-total string
    	total memory to be consumed. Memory will be consumed via multiple allocations.
  -stderrthreshold value
    	logs at or above this threshold go to stderr
  -v value
    	log level for V logs
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging